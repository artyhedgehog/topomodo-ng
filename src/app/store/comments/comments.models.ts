export enum CommentModelKey {
  TEXT = 'text'
}

export interface CommentModel {
  [CommentModelKey.TEXT]: string
}

export enum CommentsStateKey {
  ITEMS = 'items'
}

export interface CommentsState {
  [CommentsStateKey.ITEMS]: CommentModel[]
}
