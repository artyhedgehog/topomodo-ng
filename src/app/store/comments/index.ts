export * from './comments.actions'
export * from './comments.models'
export * from './comments.reducer'
export * from './comments.selectors'
