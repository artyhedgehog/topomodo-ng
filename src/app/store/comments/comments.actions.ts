import { Action } from '@ngrx/store'

import { CommentModel } from './comments.models'

export enum CommentsActionType {
  ADD = '[Comments] Add'
}

export interface AddCommentPayload {
  data: CommentModel
}

export class AddComment implements Action {
  readonly type = CommentsActionType.ADD

  static of(payload: AddCommentPayload): AddComment {
    return new AddComment(payload)
  }

  protected constructor(public readonly payload: AddCommentPayload) {}
}


