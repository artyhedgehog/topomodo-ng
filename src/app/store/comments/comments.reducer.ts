import { Action } from '@ngrx/store'
import * as R from 'ramda'

import { CommentModel, CommentsState, CommentsStateKey } from './comments.models'
import { AddComment, CommentsActionType } from './comments.actions'

const initialState: CommentsState = {
  [CommentsStateKey.ITEMS]: []
}

function addComment(state: CommentsState, action: AddComment) {
  const appendComment = R.append(action.payload.data)
  const itemsLens = R.lensProp(CommentsStateKey.ITEMS)
  const items = R.view<CommentsState, CommentModel[]>(itemsLens, state)

  return R.set(itemsLens, appendComment(items), state)
}

export function commentsReducer(state: CommentsState = initialState, action: Action = { type: null }): CommentsState {
  switch (action.type) {
    case CommentsActionType.ADD: return addComment(state, action as AddComment)
    default: return state
  }
}
