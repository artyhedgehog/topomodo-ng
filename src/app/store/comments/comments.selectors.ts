import * as R from 'ramda'

import { AppState, StoreFeature } from '../store.model'
import { CommentModel, CommentsState, CommentsStateKey } from './comments.models'
import { createFeatureSelector, createSelector } from '@ngrx/store'

export const getFeature = createFeatureSelector<CommentsState>(StoreFeature.COMMENTS)

export const getComments = createSelector<AppState, CommentsState, CommentModel[]>(
  getFeature,
  R.propOr<CommentModel[]>([], CommentsStateKey.ITEMS)
)
