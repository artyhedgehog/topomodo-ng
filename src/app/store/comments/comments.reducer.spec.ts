import { commentsReducer } from './comments.reducer'
import { CommentModel, CommentModelKey, CommentsState, CommentsStateKey } from './comments.models'
import { AddComment } from './comments.actions'

describe('commentsReducer', () => {
  it('provides default state', () => {
    expect(commentsReducer()).toHaveProperty(CommentsStateKey.ITEMS)
  })

  it('returns unmodified state for no action', () => {
    const state: CommentsState = {
      [CommentsStateKey.ITEMS]: [
        { text: 'A comment' },
        { text: 'One more comment' }
      ]
    }
    const resultState = commentsReducer(state)

    expect(resultState).toBe(state)
    expect(resultState).toMatchSnapshot()
  })

  it('returns unmodified state for unknown action', () => {
    const state: CommentsState = {
      [CommentsStateKey.ITEMS]: [
        { text: 'A comment' },
        { text: 'One more comment' }
      ]
    }
    const resultState = commentsReducer(state, { type: 'Definitely unknown action' })

    expect(resultState).toBe(state)
    expect(resultState).toMatchSnapshot()
  })

  describe('on AddComment', () => {
    let originalComments: CommentModel[]
    let commentToAdd: CommentModel
    let action: AddComment
    let originalState: CommentsState
    let resultState: CommentsState

    beforeEach(() => {
      commentToAdd = { [CommentModelKey.TEXT]: 'New added comment' }
      action = AddComment.of({ data: commentToAdd })
      originalComments = [
        { [CommentModelKey.TEXT]: 'A comment' },
        { [CommentModelKey.TEXT]: 'One more comment' },
      ]
      originalState = {
        [CommentsStateKey.ITEMS]: originalComments
      }
      resultState = commentsReducer(originalState, action)
    })

    it('leaves old state unchanged', () => {
      expect(originalState).toMatchSnapshot()
    })

    it('adds a comment to result state', () => {
      expect(resultState).toHaveProperty([CommentsStateKey.ITEMS], [
        ...originalComments,
        commentToAdd
      ])
    })
  })
})
