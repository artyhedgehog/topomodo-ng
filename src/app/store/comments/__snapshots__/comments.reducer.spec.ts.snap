// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commentsReducer on AddComment leaves old state unchanged 1`] = `
Object {
  "items": Array [
    Object {
      "text": "A comment",
    },
    Object {
      "text": "One more comment",
    },
  ],
}
`;

exports[`commentsReducer returns unmodified state for no action 1`] = `
Object {
  "items": Array [
    Object {
      "text": "A comment",
    },
    Object {
      "text": "One more comment",
    },
  ],
}
`;

exports[`commentsReducer returns unmodified state for unknown action 1`] = `
Object {
  "items": Array [
    Object {
      "text": "A comment",
    },
    Object {
      "text": "One more comment",
    },
  ],
}
`;
